Wrong position of inserted `&self` in suggestion of compile error

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
struct X(i32);

impl X {
    pub(crate) fn f() {
        self.0
    }
}

fn main() {}
```

I expected to see this happen: Correct compile error due to lack of `self` receiver.

Instead, this happened: Compile error happens. But its suggestion is not correct:

```
error[E0424]: expected value, found module `self`
 --> bar.rs:5:9
  |
4 |     pub(crate) fn f() {
  |                   - this function doesn't have a `self` parameter
5 |         self.0
  |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
  |
help: add a `self` receiver parameter to make the associated `fn` a method
  |
4 |     pub(&selfcrate) fn f() {
  |         +++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0424`.
```

`&self` is inserted to just after first `(`. But in this case, the first `(` is not for method parameter. So the position is not correct.

```
4 |     pub(&selfcrate) fn f() {
```

This line should be:

```
4 |     pub(crate) fn f(&self) {
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-apple-darwin
release: 1.57.0
LLVM version: 13.0.0
```



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"bugadani"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -298,11 +298,16 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                             .get(0)
                             .map(|p| (p.span.shrink_to_lo(), "&self, "))
                             .unwrap_or_else(|| {
+                                // Try to look for the "(" after the function name, if possible.
+                                // This avoids placing the suggestion into the visibility specifier.
+                                let span = fn_kind
+                                    .ident()
+                                    .map_or(*span, |ident| span.with_lo(ident.span.hi()));
                                 (
                                     self.r
                                         .session
                                         .source_map()
-                                        .span_through_char(*span, '(')
+                                        .span_through_char(span, '(')
                                         .shrink_to_hi(),
                                     "&self",
                                 )

--- a/src/test/ui/suggestions/suggest-add-self.rs
+++ b/src/test/ui/suggestions/suggest-add-self.rs
@@ -1,0 +1,15 @@
+struct X(i32);
+
+impl X {
+    pub(crate) fn f() {
+        self.0
+        //~^ ERROR expected value, found module `self`
+    }
+
+    pub fn g() {
+        self.0
+        //~^ ERROR expected value, found module `self`
+    }
+}
+
+fn main() {}

--- a/src/test/ui/suggestions/suggest-add-self.stderr
+++ b/src/test/ui/suggestions/suggest-add-self.stderr
@@ -1,0 +1,29 @@
+error[E0424]: expected value, found module `self`
+  --> $DIR/suggest-add-self.rs:5:9
+   |
+LL |     pub(crate) fn f() {
+   |                   - this function doesn't have a `self` parameter
+LL |         self.0
+   |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
+   |
+help: add a `self` receiver parameter to make the associated `fn` a method
+   |
+LL |     pub(crate) fn f(&self) {
+   |                     +++++
+
+error[E0424]: expected value, found module `self`
+  --> $DIR/suggest-add-self.rs:10:9
+   |
+LL |     pub fn g() {
+   |            - this function doesn't have a `self` parameter
+LL |         self.0
+   |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
+   |
+help: add a `self` receiver parameter to make the associated `fn` a method
+   |
+LL |     pub fn g(&self) {
+   |              +++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0424`.

